### PR TITLE
Adding System.Buffers support to the System.IO.Compression namespace

### PR DESCRIPTION
**Pull request by sokket**
Changing the System.IO.Compression namespace to utilize the System.Buffers implementation, which allows for pooling of byte[] and char[] buffers, which should reduce GC pressure and overhead since the buffers will reused instead of allocated on-demand each time. 

This should fix issues #4875 and #1991

/cc @stephentoub 
/fyi: @KrzysztofCwalina

### CODE REVIEW

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 1]
```diff
@@ -623,8 +624,10 @@ internal Boolean LoadLocalHeaderExtraFieldAndCompressedBytesIfNeeded()
                 _compressedBytes = new Byte[(_compressedSize / MaxSingleBufferSize) + 1][];
                 for (int i = 0; i < _compressedBytes.Length - 1; i++)
                 {
-                    _compressedBytes[i] = new Byte[MaxSingleBufferSize];
+                    _compressedBytes[i] = ArrayPool<byte>.Shared.Rent(MaxSingleBufferSize);
```
Comment (KrzysztofCwalina) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Should we add a finalizer to this type to ensure the buffers are returned even if the archive is not closed correctly?

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Adding a finalizer itself has cost, even if the instance is disposed and finalization suppressed.  But the type doesn't even implement IDisposable, so no code today is going to be disposing of it, which means we'd likely end up leaving a lot of instances to be cleaned up during finalization, which is relatively expensive.

Might be better to just not use the pool here, if we're concerned about enough of the buffers not being returned.

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs [hunk 2]
```diff
@@ -504,38 +505,45 @@ public static void ExtractToDirectory(String sourceArchiveFileName, String desti
                 // to be greater than the length of typical entry names from the file system, even
                 // on non-Windows platforms. The capacity will be increased, if needed.
                 const int DefaultCapacity = 260;
-                char[] entryNameBuffer = new char[DefaultCapacity];
+                char[] entryNameBuffer = ArrayPool<char>.Shared.Rent(DefaultCapacity);
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs]: This is being passed by ref to EntryFromPath, which passes it by ref to EnsureCapacity, which may throw it away and allocate a different buffer.  Isn't that potentially an issue, that this buffer isn't being returned and that a different one may be?

File: src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs [hunk 3]
```diff
@@ -531,6 +532,7 @@ protected override void Dispose(bool disposing)
                     }
                     finally
                     {
+                        ArrayPool<byte>.Shared.Return(_buffer);
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs]: What happens if Dispose is called multiple times?  Will _buffer be returned multiple times?

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs]: Should we be setting clearArray to true?  We're returning buffers that could potentially have been storing data that shouldn't be exposed later to others using the pool.

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Yeah, we would need to add Dispose and the compression components would have to call it.

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: > we would need to add Dispose

That would need to be added to the full framework as well.  And this is a public type, which means code beyond these compression types would now be using a finalizable object without disposing it.

My $.02: I would suggest either we don't use the pool at all, or if we think the majority of usage will hit the code paths that call Return, we accept that some buffers may not be returned.

Comment (KrzysztofCwalina) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Possibly, it's the best option unfortunately, but I do think we need to make buffer returns 100% reliable in our code. Otherwise it will give bad rep to the pooling :-)

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: >  I do think we need to make buffer returns 100% reliable in our code

I agree on 99% ;)  But I'm not sure that perfectly returning all the buffers all the time is worth the code complexity and potentially perf impact on the common/success path.  If you look at StringBuilderCache in mscorlib, for example, usage of it doesn't spend any effort to try to put back the builder on error paths... if an exception is thrown and the builder isn't put back, a new one is simply created the next time around.  That'll happen as well with ArrayPool, and with the infrequency and relative cost of exceptions, plus the code complexity added by using try/finallies, and potentially the perf impact from doing so (even minor things like functions not being inlinable where they were before by now having try/finally) or in this case by, gulp, adding finalizers, I really question whether it's worth it to achieve 100%.

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs (hunk 2 above)
Comment (sokket) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs]: Ah good catch, I missed EnsureCapacity. I'll change that function to return the buffer and rent a new, bigger one if necessary

File: src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs (hunk 3 above)
Comment (sokket) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs]: Good catch, added a guard to make sure it is only Disposed once and added ```clearArray: true```

**Comment by sokket** [comment]
@stephentoub I can't repro this failure on my box...does anything pop out at you / have you seen these build failures before when adding a new nuget package to a project?

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 4]
```diff
@@ -1064,7 +1067,12 @@ private void UnloadStreams()
         {
             if (_storedUncompressedData != null)
                 _storedUncompressedData.Dispose();
-            _compressedBytes = null;
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Why was this removed?  What if UnloadStreams is called twice?

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Am I reading the code correctly that MaxSingleBufferSize is ~2GB?  That's not going to even come from the pool, so isn't all of this just going to make things worse?

This leads me to ask: how have you verified that the changes you've made are a net improvement, are using the pool as you'd expect, etc.?

Comment (sokket) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: That looks to be correct...the original buffer pool had a max size of 2GB and I moved it down arbitrarily; however, if we have a use case for it actually being 2GB then we should move it back up to leverage that scenario. We didn't have data to make a good estimate of what the max buffer size should be, so I feel like moving it up to help this scenario would be beneficial. 

I'm working on getting those numbers now, I put this PR out to get test coverage since the Private PR run seems to have disappeared from Jenkins :(

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: > so I feel like moving it up to help this scenario would be beneficial.

I don't agree.  This code is used when unzipping files larger than 2GB-ish.  It should be relatively rare to unzip files that large, and yet the moment a process does, it's going to end up holding onto a 2GB array for the lifetime of the process.  If it's a 10GB zip file, we'll end up holding onto 5 2GB-ish arrays for the lifetime of the process.

This doesn't seem like a good trade-off to me.  If we had built-in support for reclaiming arrays in the pool after some period of time, then maybe, but as it currently stands, the pool is effectively a memory leak.

Comment (vancem) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Two comments

I don't think we should be making classes Disposable so that pooling is 100% effective in reclaiming memory.    Ideally only things that have SEMANTIC state change on death (e.g. files close themselves), should be Disposable.  I think it is OK to leak from the pool on Exceptions (since raising exceptions is supposed to be rare).   I believe this is Stephen's view as well. 

We really should have some mechanism where ArrayPool will release things from the pool if they have not been used (although it is hard to be perfect here, but we should probably release 'old' things from the pool that have not be used before a Gen2 GC (I believe we have an event).  

I also agree with Stephen that we should 'not make things worse'  in particular if we pool large arrays (e.g. > 1MB), we need some assurance that these will bel cleaned up if they are not used.

Comment (sokket) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Good points :) I'll revert my changes to this file

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 4 above)
Comment (sokket) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Will be reverted

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs [hunk 5]
```diff
@@ -504,38 +505,45 @@ public static void ExtractToDirectory(String sourceArchiveFileName, String desti
...
-                // If no entries create an empty root directory entry:
-                if (includeBaseDirectory && directoryIsEmpty)
-                    archive.CreateEntry(EntryFromPath(di.Name, 0, di.Name.Length, ref entryNameBuffer, appendPathSeparator: true));
+                    // If no entries create an empty root directory entry:
+                    if (includeBaseDirectory && directoryIsEmpty)
+                        archive.CreateEntry(EntryFromPath(di.Name, 0, di.Name.Length, ref entryNameBuffer, appendPathSeparator: true));
+                }
+                finally
+                {
+                    ArrayPool<char>.Shared.Return(entryNameBuffer);
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs]: Should we be clearing everywhere we Return?

Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs]: (I actually wonder if we have the wrong default on clearArray.   Maybe the default should be the "safe" option, namely clearing, and folks can override that to not clear to get the perf win?)

File: src/System.IO.Compression.ZipFile/tests/project.json [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {
   "dependencies": {
+    "System.Buffers" : "4.0.0-rc2-23616",
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/tests/project.json]: the rest of this file uses 23712 rather than 23616

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs (hunk 5 above)
Comment (sokket) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs]: Is clearing here necessary? I don't think the ZipFile here supports password-protected Zip files, so the uncompressed chunks shouldn't contain sensitive information...in theory

File: src/System.IO.Compression.ZipFile/tests/project.json (hunk 6 above)
Comment (sokket) [inline src/System.IO.Compression.ZipFile/tests/project.json]: Rebasing conflict :)

**Comment by sokket** [comment]
@dotnet-bot test this please

**Comment by sokket** [comment]
@stephentoub @KrzysztofCwalina any further comments?

**Comment by stephentoub** [comment]
LGTM

**Comment by sokket** [comment]
Thanks! Merging now
